src/csg.cpp

```diff
@@ -0,0 +1,174 @@
+        for (int i = 0; i < 2; i ++){
+            elements[i] = tree_data[j]["elements"][i].get<int>();
+        }
+        if (type == "Intersection"){
+            tree[j] = {Intersection, elements};
+        }else if (type == "Union"){
+            tree[j] = {Union, elements};
+        }else if (type == "Negation"){
+            tree[j] = {Negation, elements};
+        }
+    }
+    return true;
+}
+
+std::pair<std::array<double, 2>, std::vector<int>> iterTree(const std::vector<csg_unit>csgTree,const int curNode,const std::vector<std::array<double , 2>> funcInt){
+    csg_unit curUnit = csgTree[curNode - 1];
+    std::array<double, 2> interval, childInt1, childInt2;
+    std::vector<int> af(funcInt.size(), 1), childAF1(funcInt.size(), 1), childAF2(funcInt.size(), 1);
+    if (curUnit.elements[0] > 0){
+        std::pair<std::array<double, 2>, std::vector<int>> child1 = iterTree(csgTree, curUnit.elements[0], funcInt);
+        childInt1 = child1.first;
+        childAF1 = child1.second;
+    }else{
+        childInt1 = funcInt[-curUnit.elements[0] - 1];
+        childAF1[-curUnit.elements[0] - 1] = 0;
+    }
+    if (childInt1[0] * childInt1[1]>0){
+        for (size_t i = 0; i < childAF1.size(); i++){
+            childAF1[i] = 1;
+        }
+    }
+    if (curUnit.operation != Negation){
+        if (curUnit.elements[1] > 0){
+            std::pair<std::array<double, 2>, std::vector<int>> child2 = iterTree(csgTree, curUnit.elements[1], funcInt);
+            childInt2 = child2.first;
+            childAF2 = child2.second;
+        }else{
+            childInt2 = funcInt[-curUnit.elements[1] - 1];
+            childAF2[-curUnit.elements[1] - 1] = 0;
+        }
+    }
+    if (childInt2[0] * childInt2[1]>0){
+        for (size_t i = 0; i < childAF2.size(); i++){
+            childAF2[i] = 1;
+        }
+    }
+    switch (curUnit.operation){
+        case Intersection:
+            interval = {std::max(childInt1[0], childInt2[0]), std::max(childInt1[1], childInt2[1])};
+            if(interval[0]*interval[1] <= 0){
+                for (int i = 0; i < funcInt.size(); i++){
+                    af[i] = childAF1[i] * childAF2[i];
+                }
+            }
+            break;
+        case Union:
+            interval = {std::min(childInt1[0], childInt2[0]), std::min(childInt1[1], childInt2[1])};
+            if(interval[0]*interval[1] <= 0){
+                for (int i = 0; i < funcInt.size(); i++){
+                    af[i] = childAF1[i] * childAF2[i];
+                }
+            }
+            break;
+        case Negation:
+            interval = {std::min(-childInt1[0], -childInt1[1]), std::max(-childInt1[0], -childInt1[1])};
+            if(interval[0]*interval[1] <= 0)
+                af = childAF1;
+            break;
+        default:
+            std::cout << "not a valid CSG operation" << std::endl;
+    }
+    return std::pair(interval, af);
+}
+
+void prune_data(const std::vector<std::array<double, 3>>& mesh_pts,
```

Comment:
Here, the pruning of the previous meta data happens. Only the patches, chains, and non_manifold_edges_of_vert are changed such that if they are not on the CSG shape, they will be removed. All the other meta data remain unchanged.